### PR TITLE
Implement support for models in the Row of a GridLayout

### PR DESCRIPTION
**Pull request by dfaure-kdab**
This is only part of the overall if/for support in grid: for now it's only supported to repeat cells in a given row, one cannot repeat entire Rows this way yet.

Some implementation notes:
- GridLayoutOrganizedData (4 u16s per cell) can now contain indirection indices in the case of repeaters, like LayoutCache (2 f32s per cell). The "generator" for that is separate code due to these two differences (amount and type), while LayoutCacheAccess can be used for both (it now receives the additional information of how many entries per item, as a parameter)
- BoxLayoutFunction and BoxLayoutCellData are used by the grid layouting code too; good for code sharing, but bad naming then. There are TODOs to rename those in a followup commit. This code reuse required porting from LayoutInfo to BoxLayoutCellData (which has only one field, the LayoutInfo, I suppose this is good for future extensions...).
- Compared to handling of repeaters for box layouts, there are additional complications due to the "new_row" bool, especially with empty repeaters (the first item at compile time isn't necessarily the first item at runtime). The generated code solves that with a "running" new_row bool, to detect if the repeater had instances or not.

### CODE REVIEW

**Comment by dfaure-kdab** [comment]
Known issues:
 - specifying a row,col,rowspan or colspan in a repeated element leads to `Repeated element should be empty because of the repeater_component.rs pass` (elem.debug[0].layout is none at that point, when visiting all named references from `repeater_components::adjust_references`, because this happens before lower_layout...)
 - no documentation/example yet. Hmm, actually the `if/for` (model) syntax isn't documented currently in HorizontalLayout and VerticalLayout, only in the tiles example?

**Comment by dfaure-kdab** [comment]
Somewhat real-world example (at least more than the auto test):
[MainWindow.slint.txt](https://github.com/user-attachments/files/24099589/MainWindow.slint.txt)
Screenshot_20251211_105229

File: tests/cases/layout/grid_with_model_in_row.slint [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    out property <bool> test_after_clicked: end.row == 2 && end.col == 1 && clicked && before_cells_ok;
+}
+
+/*
+
+```cpp
+auto handle = TestCase::create();
+const TestCase &instance = *handle;
+assert(instance.get_test());
+```
```
Comment (ogoffart) [inline tests/cases/layout/grid_with_model_in_row.slint:102]: Would be nice to also test the C++ code generator like you do with rust.
I think you can pretty much copy/paste the rust code and just remove the `!` for the assert.

The JS is a bit more complicated that just copy paste, but very similar nonetheless

File: api/cpp/include/slint.h [hunk 2]
```diff
@@ -176,9 +185,17 @@ box_layout_info_ortho(cbindgen_private::Slice<cbindgen_private::BoxLayoutCellDat
 }
 
 /// Access the layout cache of an item within a repeater
-inline float layout_cache_access(const SharedVector<float> &cache, int offset, int repeater_index)
+inline float layout_cache_access(const SharedVector<float> &cache, int offset, int repeater_index,
+                                 int entries_per_item)
```
Comment (ogoffart) [inline api/cpp/include/slint.h]: Candidate for a template function?
```suggestion
template<typename T>
inline T layout_cache_access(const SharedVector<T> &cache, int offset, int repeater_index,
                                 int entries_per_item)
```

On the other hand, since it is so small, not sure it is worth it

File: internal/compiler/expression_tree.rs [hunk 3]
```diff
@@ -757,6 +757,7 @@ pub enum Expression {
         /// When set, this is the index within a repeater, and the index is then the location of another offset.
         /// So this looks like `layout_cache_prop[layout_cache_prop[index] + repeater_index]`
         repeater_index: Option<Box<Expression>>,
+        entries_per_item: usize,
```
Comment (ogoffart) [inline internal/compiler/expression_tree.rs:764]: Could probably benefit from a doc comment.  Also, is the formula two line above need an update?

File: internal/compiler/expression_tree.rs [hunk 4]
```diff
@@ -1881,7 +1882,7 @@ pub fn pretty_print(f: &mut dyn std::fmt::Write, expression: &Expression) -> std
             write!(f, "return ")?;
             e.as_ref().map(|e| pretty_print(f, e)).unwrap_or(Ok(()))
         }
-        Expression::LayoutCacheAccess { layout_cache_prop, index, repeater_index } => {
+        Expression::LayoutCacheAccess { layout_cache_prop, index, repeater_index, .. } => {
```
Comment (ogoffart) [inline internal/compiler/expression_tree.rs]: This would probably help future debugging to include the `entries_per_item`

File: internal/compiler/llr/expression.rs [hunk 5]
```diff
@@ -171,13 +171,14 @@ pub enum Expression {
...
-        /// So this looks like `layout_cache_prop[layout_cache_prop[index] + repeater_index]`
+        /// So this looks like `layout_cache_prop[layout_cache_prop[index] + repeater_index * entries_per_item]`
         repeater_index: Option<Box<Expression>>,
+        entries_per_item: usize,
     },
     /// Will call the sub_expression, with the cell variable set to the
     /// array of BoxLayoutCellData from the elements
     BoxLayoutFunction {
-        /// The local variable (as read with [`Self::ReadLocalVariable`]) that contains the sell
+        /// The local variable (as read with [`Self::ReadLocalVariable`]) that contains the cell
```
Comment (ogoffart) [inline internal/compiler/llr/expression.rs]: Is it maybe `cells` plurral?   (same problem bellow as well)

File: internal/compiler/generator/rust.rs [hunk 6]
```diff
@@ -3420,6 +3431,81 @@ fn struct_name_to_tokens(name: &StructName) -> Option<proc_macro2::TokenStream>
     }
 }
 
+fn grid_input_function(
```
Comment (ogoffart) [inline internal/compiler/generator/rust.rs:3466]: I guess you did consider trying to merge this with `box_layout_function`, and that wasn't worth it?

File: internal/compiler/tests/syntax/layout/if_in_grid.slint
Comment (ogoffart) [inline internal/compiler/tests/syntax/layout/if_in_grid.slint:1]: Why did you delete this file. We should still get some errors for it so i believe it should stay.

File: internal/core/model.rs [hunk 7]
```diff
@@ -821,7 +821,8 @@ pub trait RepeatedItemTree:
         LogicalLength::default()
     }
 
-    /// Returns what's needed to perform the layout if this ItemTrees is in a box layout
+    /// Returns what's needed to perform the layout if this ItemTree is in a layout
+    // PENDING(dfaure) rename to layout_cell_info, this is also used by grid layouts
```
Comment (ogoffart) [inline internal/core/model.rs:825]: yes, that's a better name.

File: tests/cases/layout/grid_with_model_in_row.slint (hunk 1 above)
Comment (dfaure-kdab) [reply tests/cases/layout/grid_with_model_in_row.slint:102]: Ooops, I forgot to do that indeed. Will do.

File: internal/compiler/tests/syntax/layout/if_in_grid.slint
Comment (dfaure-kdab) [reply internal/compiler/tests/syntax/layout/if_in_grid.slint:1]: No, there are no errors anymore in that file. if/for are supported when used directly under a GridLayout.

The remaining error is the one in `if_in_grid_row.slint`.

File: internal/core/model.rs (hunk 7 above)
Comment (dfaure-kdab) [reply internal/core/model.rs:825]: OK, I will do the renamings in a followup MR, to make review easier.

**Comment by dfaure-kdab** [comment]
Shall I add the example in slint/examples?

I'm wondering if it wouldn't be more useful if the model came from some rust code rather than being hardcoded in the slint file? (but then C++ users wouldn't find it that great...)

File: api/cpp/include/slint.h (hunk 2 above)
Comment (dfaure-kdab) [reply api/cpp/include/slint.h]: Done, it will be useful if we ever need a third data type ;)

File: internal/compiler/llr/expression.rs (hunk 5 above)
Comment (dfaure-kdab) [reply internal/compiler/llr/expression.rs]: indeed, fixed.

File: internal/compiler/generator/rust.rs (hunk 6 above)
Comment (dfaure-kdab) [reply internal/compiler/generator/rust.rs:3466]: You guess right ;)
* The input arguments differ
** box layout wants an orientation, grid doesn't
** for each repeater, boxlayout only needs the repeater number, grid also needs the new_row value (and I'm guessing we'll need more information for the next feature, repeating Rows), which is why the second type in the `Either` has changed
* the code being pushed for each instance is different obviously, including a variable outside the for loop, for grids -- I'm not sure how this could be passed as parameter given that stuff inside `quote!` refers to local variables outside of it, coming from the shared logic

Maybe instead of a single function with a ton of parameters, there could be shared helpers between the two? The last 13 lines are identical, for instance. I can look into factorizing that if you think it's a good idea (however it'll be a function with 7 variables, AFAICS: repeater_idx, cells_variable, sub_expression, fixed_count, repeated_count, push_code)

Or maybe this is a job for a macro? It can provide the beginning and end, and let the two functions implement the inner stuff differently. But the difference in type for `repeater` makes this tricky.

File: internal/compiler/expression_tree.rs (hunk 3 above)
Comment (dfaure-kdab) [reply internal/compiler/expression_tree.rs:764]: Ah, yes, I updated the formula in expression.rs but not in expression_tree.rs. Removed the duplication and updated  the comments.

File: internal/compiler/expression_tree.rs (hunk 4 above)
Comment (dfaure-kdab) [reply internal/compiler/expression_tree.rs]: Hmm. The current output is
```
   for person[] in (root.model/* as model */):(base)  := Cell {  /* Cell,;Rectangle,Cell::root */
      cell-text: .@model.account;
      col: <borrowed>.layout-organized-data[12 + $index];
      height: <borrowed>.layout-cache-v[7 + $index];
      row: <borrowed>.layout-organized-data[14 + $index];
      width: <borrowed>.layout-cache-h[7 + $index];
      [...]
   }
```
which is actually incorrect, we don't just do 12 + $index. I never noticed, I didn't use this for debugging...
Here's a fixed version. More correct, but certainly less readable?
```
   for person[] in (root.model/* as model */):(base)  := Cell {  /* Cell,;Rectangle,Cell::root */
      cell-text: .@model.account;
      col: <borrowed>.layout-organized-data[<borrowed>.layout-organized-data[12] + $repeater_index * 4];
      height: <borrowed>.layout-cache-v[<borrowed>.layout-cache-v[7] + $repeater_index * 2];
      row: <borrowed>.layout-organized-data[<borrowed>.layout-organized-data[14] + $repeater_index * 4];
      width: <borrowed>.layout-cache-h[<borrowed>.layout-cache-h[7] + $repeater_index * 2];
```

**Comment by dfaure-kdab** [comment]
> > specifying a row,col,rowspan or colspan in a repeated element leads to Repeated element should be empty because of the repeater_component.rs pass
> 
> Looks like we should put these properties not within the repeated component, but within the base of the repeated component (in the layout pass), which is the thing actually being repeated.

There are multiple issues. I fixed the first few, but I could use your help with what I hope is the last one.

* The error above actually came from materialize_fake_properties() adding colspan to the repeater, I simply added a `if elem.repeated.is_some() { return; }` there.
* Then I realized what's missing is to actually use the 4 properties when generating code in grid_input_function. I added them to GridLayoutRepeatedElement. But: now I hit the unreachable in `map_property_reference` because it's a property on a LoweredElement::Repeated element. I think that's because it started being used (by lower_layout) *after* the code in repeater_components that adjusts references to properties in repeated elements. So my idea was to do a similar adjustment in lower_layout.rs:
```
--- i/internal/compiler/passes/lower_layout.rs
+++ w/internal/compiler/passes/lower_layout.rs
@@ -389,7 +389,25 @@ fn add_element(
         }
 
         let propref_or_default = |name: &'static str| -> Option<RowColExpr> {
-            crate::layout::binding_reference(item_element, name).map(RowColExpr::Named)
+            let nr = crate::layout::binding_reference(item_element, name).map(|nr| {
+                // similar to adjust_references in repeater_component.rs
+                let e = nr.element();
+                let mut nr = nr.clone();
+                if e.borrow().repeated.is_some()
+                    && let crate::langtype::ElementType::Component(c) = e.borrow().base_type.clone()
+                {
+                    nr = NamedReference::new(&c.root_element, nr.name().clone())
+                };
+                nr
+            });
+            nr.map(RowColExpr::Named)
         };
``` 

This "works" but everything falls apart when this named reference is used. I hit the unwrap() in map_property_reference because `enclosing` isn't set? I don't really understand what enclosing is, but I see that it's also not set when repeater_components does similar adjustments, it gets set later on?

```
[...]
adjust_references: Adjusting reference to property height in repeated element "optionalText" to refer to component root "optionalText_repeated". Enclosing=""
grid_layout_input_data: adjusting reference to property colspan in repeated element "optionalText" to refer to component root "optionalText_repeated". Enclosing=""
[...]
```
(the first one is the existing code, the second one is my new code)
which leads to
```
Mapping property reference root-1.empty-5-layout-organized-data in component "MainWindow". Enclosing="MainWindow"
Mapping property reference root-1.empty-5-height in component "MainWindow". Enclosing="MainWindow"
Mapping property reference root-1.empty-5-layout-organized-data in component "MainWindow". Enclosing="MainWindow"
Mapping property reference optionalText_repeated-6.colspan in component "MainWindow". Enclosing=""
=> unwrap on None
```
Should I keep digging in this direction, or am I missing a much simpler way to solve this? There's some complex interaction between the various compiler passes... any hint welcome.
(I think this can all be done in a followup PR, but that's up to you)

**Comment by dfaure-kdab** [comment]
I think I understand what the problem is. The ExpressionLoweringCtx is for the MainWindow component, while the property is in the repeated component, which is fully separate? AFAIU this means it's inaccessible from there? Like the problem that we can't access properties of popups, maybe?

**Comment by ogoffart** [comment]
right, the `col`, `row` and span property need to be located in the repeated component itself, not on the placeholder element in the parent.
And their expression is dependent of the repeated instance at runtime (each instance might have a different value).
So we need a way to access it like we have `RepeatedItemTree::box_layout_data` and we should either add the cell information there, or add a `RepeatedItemTree::grid_layout_cell_data` function to get this.

**Comment by dfaure-kdab** [comment]
The CI failure look like rustc crashing / running out of memory, not related to this change.

File: internal/interpreter/eval_layout.rs [hunk 8]
```diff
@@ -221,60 +225,125 @@ fn repeater_instances(
...
+            RowColExpr::Literal(value) => *value,
+            RowColExpr::Named(nr) => {
+                let value: f32 = eval::load_property(component, &nr.element(), nr.name())
+                    .unwrap()
+                    .try_into()
+                    .unwrap();
+                if value >= 0.0 && value <= u16::MAX as f32 {
+                    value as u16
+                } else {
+                    panic!(
```
Comment (ogoffart) [inline internal/interpreter/eval_layout.rs]: We should not panic there.  Can you add a test where `col` comes from the model and is negative or way too big?

Ideally we should notify the developer that there is something wrong, but we don't have that right now (we were thinking showing that in the live preview somehow)
I think we could just saturate.

File: internal/compiler/llr/lower_to_item_tree.rs [hunk 9]
```diff
@@ -545,6 +547,15 @@ fn lower_sub_component(
         crate::layout::Orientation::Vertical,
     )
     .into();
+    grid_layout_cell_for_component(component).map(|parent_grid_layout_cell| {
```
Comment (ogoffart) [inline internal/compiler/llr/lower_to_item_tree.rs]: maybe `sub_component.grid_layout_input_for_repeated = grid_layout_cell_for_component(...).map(...);` would be more readable? (I know clippy contains about Option::map returning `()`)

File: internal/compiler/llr/lower_to_item_tree.rs [hunk 10]
```diff
@@ -605,6 +616,33 @@ fn lower_sub_component(
     LoweredSubComponent { sub_component, mapping }
 }
 
+fn grid_layout_cell_for_component(component: &Rc<Component>) -> Option<GridLayoutElement> {
```
Comment (ogoffart) [inline internal/compiler/llr/lower_to_item_tree.rs]: This is really a hack to re-use debug data for this.
I'm thinking the info from `GridLayoutElement` could be stored within the `Element` in an Option, would that not be better?

**Review by ogoffart — COMMENTED** [review]
Thanks, we're getting there.

I think this should benefit from test that have dynamic `col` and `row` that depends on the model, and also testing invalid values for col/row/colspan/rowspan  (negative value or really big)

File: internal/compiler/llr/lower_to_item_tree.rs (hunk 10 above)
Comment (dfaure-kdab) [reply internal/compiler/llr/lower_to_item_tree.rs]: Yeah, I hesitated between the two approaches (to avoid bloating up `Element`). But indeed it would be simpler and less hackish. Will do.

File: internal/compiler/llr/lower_to_item_tree.rs (hunk 9 above)
Comment (dfaure-kdab) [reply internal/compiler/llr/lower_to_item_tree.rs]: Ah, I didn't know that. If only the code was clippy clean, we could enable it in the commit hook ;)

File: internal/interpreter/eval_layout.rs (hunk 8 above)
Comment (dfaure-kdab) [reply internal/interpreter/eval_layout.rs]: Done, we now warn and saturate (at runtime, both in case of compiler and interpreter).

File: internal/compiler/llr/lower_to_item_tree.rs (hunk 10 above)
Comment (dfaure-kdab) [reply internal/compiler/llr/lower_to_item_tree.rs]: Done.

File: examples/layouts/grid-with-model-in-rows.slint
Comment (ogoffart) [inline examples/layouts/grid-with-model-in-rows.slint:1]: Can we test this? The interpreter driver has some test for examples but they need to he added manually in the main.rs
https://github.com/slint-ui/slint/blob/master/tests/driver/interpreter/main.rs#L33

That way we at least check that this compiles without errors.

File: internal/compiler/generator/rust.rs [hunk 11]
```diff
@@ -1903,13 +1930,13 @@ fn generate_repeated_component(
             }
         }
     } else {
-        // TODO: we could generate this code only if we know that this component is in a box layout
```
Comment (ogoffart) [inline internal/compiler/generator/rust.rs:1906]: (this is still valid with `s/box //`)

File: internal/compiler/layout.rs [hunk 12]
```diff
@@ -278,7 +278,7 @@ impl LayoutConstraints {
 #[derive(Debug, Clone)]
 pub enum RowColExpr {
     Named(NamedReference),
-    Literal(u16),
+    Literal(f64), // u16 range, or ROW_COL_AUTO
```
Comment (ogoffart) [inline internal/compiler/layout.rs]: How about adding `Auto` to this enum?

But I guess this makes the code generater simpler the way it is now.

File: internal/compiler/object_tree.rs [hunk 13]
```diff
@@ -873,6 +873,9 @@ pub struct Element {
     /// How many times the element was inlined
     pub inline_depth: i32,
 
+    /// Information about the grid cell containing this element, if applicable
+    pub grid_layout_cell: Option<crate::layout::GridLayoutElement>,
```
Comment (ogoffart) [inline internal/compiler/object_tree.rs]: Note that the GridLayoutElement contains a LayoutItem which is a Rc of the element itself, creating a Rc loop.
What can be done is to remove the `item` from the GridLayoutElement, and have GridLayout::elems be a Vec of LayoutItem, and get the gridlayout info from this.

Note that the expressions in grid_layout_cell also need to be visited.

File: internal/core/layout.rs [hunk 14]
```diff
@@ -439,12 +441,15 @@ pub struct GridLayoutData {
 pub struct GridLayoutInputData {
     /// whether this cell is the first one in a Row element
     pub new_row: bool,
-    /// col and row number (u16::MAX means auto).
-    pub col: u16,
-    pub row: u16,
+    /// col and row number.
+    /// Only ROW_COL_AUTO and the u16 range are valid, values outside of
+    /// that will be clamped with a warning at runtime
+    pub col: f64,
```
Comment (ogoffart) [inline internal/core/layout.rs]: I wonder if f64 and operations on them might be quite expensive 32bits MCUs 
But probably OK.

File: internal/core/layout.rs [hunk 15]
```diff
@@ -555,42 +657,90 @@ pub fn organize_dialog_button_layout(
...
     let mut organized_data = GridLayoutOrganizedData::default();
-    organized_data.reserve(input_data.len() * 4);
+    organized_data.resize(input_data.len() * 4 + repeater_indices.len() * 2, 0 as _);
+    let mut generator =
+        OrganizedDataGenerator::new(repeater_indices.as_slice(), &mut organized_data);
+    let mut errors = Vec::new();
+
+    fn clamp_to_u16(value: f64, field_name: &str, errors: &mut Vec<String>) -> u16 {
+        if value < 0.0 {
+            errors.push(format!("cell {} {} is negative, clamping to 0", field_name, value));
```
Comment (ogoffart) [inline internal/core/layout.rs]:
```suggestion
            errors.push(format!("cell {field_name} {value} is negative, clamping to 0"));
```

File: internal/core/layout.rs [hunk 16]
```diff
@@ -555,42 +657,90 @@ pub fn organize_dialog_button_layout(
...
 
+type Errors = Vec<String>;
+
+pub fn organize_grid_layout(
+    input_data: Slice<GridLayoutInputData>,
+    repeater_indices: Slice<u32>,
+) -> GridLayoutOrganizedData {
+    let (organized_data, errors) = organize_grid_layout_impl(input_data, repeater_indices);
+    for error in errors {
+        crate::debug_log!("GridLayout error: {}", error);
```
Comment (ogoffart) [inline internal/core/layout.rs]: I'm thinking it would be useful to mention this is related to Slint.
```suggestion
        crate::debug_log!("Slint layout error: {}", error);
```

(In debug mode, we could perhaps pass a `&'static str` with the location witin the Slint file, but this doesn't have to be in that PR)

File: tests/cases/layout/grid_with_x_y_model.slint [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+    width: 1000phx;
+
+    out property <string> clicked_text;
+
+    property <[{ row: int, col: int}]> xy_model: [
+        { row: 0, col: 2 },
+        { row: 0, col: 1 },
+        { row: 2, col: -1 }, // This one should warn and saturate to 2, 0
+        { row: -500, col: 0 }, // This one should warn and saturate to 0, 0
+        { row: 1, col: 0 },
```
Comment (ogoffart) [inline tests/cases/layout/grid_with_x_y_model.slint:15]: Perhaps also add one case with `col: 1000000` or one too big value.

File: internal/compiler/generator/rust.rs (hunk 11 above)
Comment (dfaure-kdab) [reply internal/compiler/generator/rust.rs:1906]: Oh right, I didn't realize that a repeated element didn't have to be in a layout (could use absolute positioning, of course). So while this applies to both box and grid layouts, the TODO remains.

File: internal/core/layout.rs (hunk 16 above)
Comment (dfaure-kdab) [reply internal/core/layout.rs]: Hmm... I see the idea (a complex Rust application might get warnings/errors from many different crates), but note that it's really not consistent right now.

```
internal/backends/winit/winitwindowadapter.rs:899:                    i_slint_core::debug_log!("Error watching for xdg color schemes: {}", err);
internal/core/graphics/image/cache.rs:92:                             crate::debug_log!("Error loading SVG from {}: {}", &path, err);
internal/core/graphics/image/cache.rs:102:                    crate::debug_log!("Error loading image from {}: {}", &path, decode_err);
internal/core/graphics/image/cache.rs:127:                            crate::debug_log!("Error loading SVG: {}", svg_err);
internal/core/graphics/image/cache.rs:150:                    crate::debug_log!("Error decoding embedded image: {}", decode_err);
internal/interpreter/eval.rs:1459:                    corelib::debug_log!("Error loading custom font {}: {}", s.as_str(), err);
```

I see a few "Slint: [...]" but really not many. I'll change it of course, this is JFYI.

File: tests/cases/layout/grid_with_x_y_model.slint (hunk 17 above)
Comment (dfaure-kdab) [reply tests/cases/layout/grid_with_x_y_model.slint:15]: Note that it's already tested in the unittest:

internal/core/layout.rs:1227:        assert_eq!(errors[1], "cell row 80000 is too large, clamping to 65535");

I initially had it here, but it's impossible to really test what happens at runtime, column 65535 is "very much to the right", but how much depends on the window width, and it's not really right-aligned, there's some empty space to the right of the last column for some reason. So instead of adding a case that we can't test, here, I preferred adding a proper unittest where we can test that the warning shows up and the column is clamped to 65535.

File: internal/core/layout.rs (hunk 14 above)
Comment (dfaure-kdab) [reply internal/core/layout.rs]: I thought you said f64 was used internally for everything anyway....

File: examples/layouts/grid-with-model-in-rows.slint
Comment (dfaure-kdab) [reply examples/layouts/grid-with-model-in-rows.slint:1]: Thanks for the tip, added.

File: internal/core/layout.rs (hunk 14 above)
Comment (ogoffart) [reply internal/core/layout.rs]: It is used in the interprerer, but not when compiled to Rust/C++

File: internal/compiler/layout.rs (hunk 12 above)
Comment (dfaure-kdab) [reply internal/compiler/layout.rs]: Great idea, it actually makes things simpler (because we can stick to u16) and clearer.

File: internal/compiler/object_tree.rs (hunk 13 above)
Comment (dfaure-kdab) [reply internal/compiler/object_tree.rs]: I'm confused. If GridLayout::elems is Vec<LayoutItem>, where are the GridLayoutElements stored?
Removing `item` from GridLayoutElement seems like a major refactoring, it's used in many many places.

The thing is, in grid_layout_cell, I don't need the LayoutItem part of GridLayoutElement. So I could just store what I need here, without using GridLayoutElement (i.e. extract a GridLayoutCell struct from GridLayoutElement)...
Done, let me know what you think.

File: internal/core/layout.rs (hunk 14 above)
Comment (dfaure-kdab) [reply internal/core/layout.rs]: ah, I see. Changed.

File: internal/compiler/layout.rs [hunk 18]
```diff
@@ -279,24 +279,30 @@ impl LayoutConstraints {
...
     pub col_expr: RowColExpr,
     pub row_expr: RowColExpr,
     pub colspan_expr: RowColExpr,
     pub rowspan_expr: RowColExpr,
+}
+
+/// An element in a GridLayout
+#[derive(Debug, Clone)]
+pub struct GridLayoutElement {
+    pub cell: GridLayoutCell,
```
Comment (ogoffart) [inline internal/compiler/layout.rs:297]: i'm thinking you can access that with `item.grid_layout_cell` and don't need to have duplicate info that have the risk to become out of sync in the different passes.

Comment (dfaure-kdab) [reply internal/compiler/layout.rs:297]: I gave it a try, and I think it makes things really complicated.
There are two Elements in play, in case of repeaters. The repeater container element and the root element inside the repeater element.
When processing the layout in grid_layout_input_data (and wondering e.g. about new_row), we're looking at the container element, while in lower_sub_component() we are looking at the root element of the repeated component to see if it has a grid_layout_cell. It gets a bit messy.

I think I can make it work by duplicating only the bool new_row, which doesn't have the same issue...

Conceptually I have to wonder if it's the right approach though. This information feels much more at home in GridLayoutElement than in Element itself, it's just that we want to access it when generating repeated components, where we only know the component's Element...
Maybe the GridLayoutCell information could be shared between GridLayoutElement and the component Element using Rc?

Comment (ogoffart) [reply internal/compiler/layout.rs:297]: > Maybe the GridLayoutCell information could be shared between GridLayoutElement and the component Element using Rc?

Seems like a good idea.

File: internal/compiler/generator/rust.rs (hunk 11 above)
Comment (dfaure-kdab) [reply internal/compiler/generator/rust.rs:1906]: I implemented this TODO in the last commit of this PR, BTW.

File: internal/core/model.rs (hunk 7 above)
Comment (dfaure-kdab) [reply internal/core/model.rs:825]: https://github.com/slint-ui/slint/pull/10374

File: internal/compiler/layout.rs (hunk 18 above)
Comment (dfaure-kdab) [reply internal/compiler/layout.rs:297]: Done in https://github.com/slint-ui/slint/pull/10375
